sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala

```diff
@@ -2157,7 +2157,7 @@ class SubquerySuite extends QueryTest
     }
   }
 
-  test("Merge non-correlated scalar subqueries from different parent plans") {
```

Comment:
Unfortunately, this is also a valid test and so it shouldn't be altered.
I don't want to block this fix so if we need it urgently I'm ok with mergint this PR but probably we can come up with a better alternative.

I think the issue is that we try merging a new plan that contains a `ScalarSubqueryReference`  that references a cached plan with the same cached plan.

Reply:
I think the issue is that currently we allow merging a subquery to another one that is needed to compute the subquery itself. So before merging we should collect the (transitive) subquery references and don't try merging the new plan to those.
@dtenedor, let me know if you want to update this PR or I can open one with the fix tomorrow.

Reply:
@peter-toth Yeah, the bug is in `tryMergePlans` where it tries to merge a subquery plan with another plan that contains the original subquery plan inside, e.g. the regression test `select(select sum((select sum(col) from t)) from t`. I tried doing that by keeping sets of visited plans and checking them before merging, but it got complex since some `ScalarSubqueryReference`s were already converted.

We should probably merge this PR to fix planning errors while keeping most of the optimizations from this rule in the short term. Then we can follow-up to restore the remaining optimization. I put a TODO for that; with the regression test present, it should be safe to proceed from there.